docs/features/source-generators.cookbook.md

```diff
@@ -0,0 +1,250 @@
+by design, cannot interact with each other, language features implemented in this way would quickly become incompatible with other additions to the language.
+
+### Code-not-source generators
+
+There are many post-processing tasks that users perform on their assemblies today, which here we define as 'code-not-source' generation. These include, but
+are not limited to:
+
+- Optimization
+- Logging injection
+- IL Weaving
+- Call site re-writing
+
+While these techniques have many valuable use cases, they do not fit into the idea of *source* generation. They are, by defintion, code altering operations
+which is explicitly ruled out by the source generation proposal.
+
+There are already well supported tools and techniques for acheiving these kinds of operations, and the source generators proposal is not aimed at replacing them.
+
+## Designs
+
+This section is broken down by user scenarios, with general solutions listed first, and more specific examples later on.
+
+### Generated class
+
+**User scenario:** As a generator author I want to be able to create a class that can be referenced from user code.
+
+**Solution:** Generate the class based on the data available in the compilation, and add the new source to compilation.
+
+**Example:**
+
+```csharp
+public class CustomGenerator : ISourceGenerator
+{
+    public void Execute(SourceGeneratorContext context)
+    {
+        context.AddSource("myGeneratedFile.cs", SourceText.From($@"
+namespace GeneratedNamespace
+{
+    public class GeneratedClass
+    {
+        public static void GeneratedMethod()
+        {
+            // generated code
+        }
+    }
+}"));
+    }
+}
+```
+
+```csharp
+public partial class UserClass
+{
+    public void UserMethod()
+    {
+        // call into a generated method
+        GeneratedNamespace.GeneratedClass.GeneratedMethod();
+    }
+}
+```
+
+### Single file transformation
+
+**User scenario:** As a generator author I want to be able to transform a single file into a C# representation.
+
+**Solution:** Use the additional files property of the `SourceGeneratorContext` to retreive the contents of the file, convert it to the C# representation and return it.
+
+**Example:**
+
+```csharp
+public class FileTransformGenerator : ISourceGenerator
+    {
+        public void Execute(SourceGeneratorContext context)
+        {
+            // find anything that matches our files
+            var myFiles = context.AnalyzerOptions.AdditionalFiles.Where(at => at.Path.EndsWith(".myextension"));
+            foreach (var file in myFiles)
+            {
+                var content = file.GetText(context.CancellationToken);
+
+                // do some transforms based on the file contenxt
+
+                string output = "namespace MyNS { public class MyClass { } }";
+                var sourceText = SourceText.From(output);
+
+                context.AddSource($"{file.Name}generated.cs", sourceText);
+            }
+        }
+    }
+```
+
+### Augment user code
```

Comment:
Should we provide a scenario that includes partial methods? Reasons:

* **Control of name**. The developer can control the name of the member
* **Generation is optional/depending on other state**. Based on other information, generator might decide that the method isn't needed.